[clang] Ignore GCC 11 `[[malloc(x)]]` attribute

aloisklink opened the pull request:
Ignore the `[[malloc(x)]]` or `[[malloc(x, 1)]]` function attribute syntax added in [GCC 11][1].

Unlike `[[malloc]]` with no arguments (which is supported by Clang), GCC uses the one or two argument form to specify a deallocator for GCC's static analyzer.

Code currently compiled with `[[malloc(x)]]` or `__attribute__((malloc(x)))` fails with the following error: `'malloc' attribute takes no arguments`.

[1]: https://gcc.gnu.org/git/?p=gcc.git;a=commitdiff;f=gcc/doc/extend.texi;h=dce6c58db87ebf7f4477bd3126228e73e4eeee97#patch6

Fixes: https://github.com/llvm/llvm-project/issues/51607
Partial-Bug: https://github.com/llvm/llvm-project/issues/53152 (this PR only ignores `__attribute__((malloc(x)))` and allows it to compile, so only partially fixes the problem).

---

In the future, we can add this attribute to the AST as well. This will let us improve the clang static analyzer's [`unix.MismatchedDeallocator`](https://clang.llvm.org/docs/analyzer/checkers.html#id93) checker.

aloisklink committed:
[clang] Ignore GCC 11 [[malloc(x)]] attribute

Ignore the `[[malloc(x)]]` or `[[malloc(x, 1)]]` function attribute
syntax added in [GCC 11][1] and print a warning instead of an error.

Unlike `[[malloc]]` with no arguments (which is supported by Clang),
GCC uses the one or two argument form to specify a deallocator for
GCC's static analyzer.

Code currently compiled with `[[malloc(x)]]` or
`__attribute((malloc(x)))` fails with the following error:
`'malloc' attribute takes no arguments`.

[1]: https://gcc.gnu.org/git/?p=gcc.git;a=commitdiff;f=gcc/doc/extend.texi;h=dce6c58db87ebf7f4477bd3126228e73e4eeee97#patch6

Fixes: https://github.com/llvm/llvm-project/issues/51607
Partial-Bug: https://github.com/llvm/llvm-project/issues/53152

tbaederr commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -4122,6 +4122,9 @@ def RestrictDocs : Documentation {
 storage for any other object accessible to the caller.
+
+The form of ``malloc`` with one or two arguments (supported by GCC 11) is
+currently ignored by Clang.
```
Is it really ignored or is it just treated as if it had no arguments?

aloisklink replied:
In this PR, it's completely ignored, it's not even added to the Clang AST.

It just reaches the end of this function where nothing happens (let me know if I can improve the comment!):

https://github.com/llvm/llvm-project/blob/f9c914729a5f5ac7f8b61ea2d39509ff0236a228/clang/lib/Sema/SemaDeclAttr.cpp#L2084-L2087

We can't treat it as if has no arguments because in GCC it means two different things:
  - `__attribute__((malloc))` with no args means the return value is guaranteed not to alias to any other pointer (aka like [`__declspec(restrict)` in MSVC](https://learn.microsoft.com/en-us/cpp/cpp/restrict?view=msvc-170)) and that the function will normally return non-`NULL`, allowing optimizations,
  - `__attribute__((malloc(deallocator)))` instead says the returned pointer should be deallocated by the specified deallocator, allowing a static analyzer to print warnings.

See the `malloc` section of https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html for more info.

To be honest, I feel like GCC should have given the one/two argument form a different attribute name to make things less confusing, but GCC 11 has already been out for years unfortunately.

erichkeane replied:
I would like it if we updated the documentation here to be something more like:

```
The ``malloc`` attribute has two forms with different functionality. The first
is when it is used without arguments, where it marks that a function acts like
a system memory allocation function, returning a pointer to allocated storage
that does not alias storage from any other object accessible to the caller.

The second form is when ``malloc`` takes one or two arguments. The first argument
names a function that should be associated with this function as its deallocation 
function. When this form is used, it enables the compiler to diagnose when the
incorrect deallocation function is used with this variable. However the associated
warning, spelled `-Wmismatched-dealloc` in GCC, is not yet implemented in clang.

```

erichkeane commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -2064,13 +2064,26 @@ static void handleTLSModelAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+    return;
+  }
+
+  // FIXME: GCC uses [[malloc(my_func)]] to specify a deallocator for the
```
I think I'd like this to be diagnosed.  We shouldn't be silently ignoring the attribute.

aloisklink replied:
I slightly disagree, since even if we do add support for this attribute, only the Clang analyzer would use it, but I've added a `'malloc' attribute ignored because Clang does not support the one/two argument form` warning in commit https://github.com/llvm/llvm-project/commit/a76561f522f628b0882572f8dabee6f7e4abd5f5.

AaronBallman reviewed (commented):
Thank you for the fix!

We generally want ignored attributes to be diagnosed as being ignored; otherwise users have a much harder time determining whether the attribute is working or not. I think we should issue an "attribute ignored" warning when we're dropping the attribute in the AST.

Also, the changes should come with a release note so that users know about the bug fix.

aloisklink commented:
> We generally want ignored attributes to be diagnosed as being ignored; otherwise users have a much harder time determining whether the attribute is working or not. I think we should issue an "attribute ignored" warning when we're dropping the attribute in the AST.

Personally, I think there shouldn't be a warning, because even if we add the attribute to the AST, Clang will do nothing with it; the attribute would only be used by Clang's static analyzer. But I've added a warning anyway in https://github.com/llvm/llvm-project/commit/a76561f522f628b0882572f8dabee6f7e4abd5f5.

I think a generic `'malloc' attribute ignored` warning would have been pretty confusing, (especially because most of the time, people use both the no-argument form and the one/two argument form on the same function, e.g. `__attribute__((malloc, malloc(my_cleanup))) void *my_func();`) so instead I've made a custom warning that says: `'malloc' attribute ignored because Clang does not support the one/two argument form`.

> Also, the changes should come with a release note so that users know about the bug fix.

:+1: That makes sense! I've added a note in the **Bug Fixes to Attribute Support** section in https://github.com/llvm/llvm-project/commit/a76561f522f628b0882572f8dabee6f7e4abd5f5.

shafik commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -1629,6 +1629,8 @@ def IFunc : Attr, TargetSpecificAttr<TargetELF> {
 
 def Restrict : InheritableAttr {
   let Spellings = [Declspec<"restrict">, GCC<"malloc">];
+  let Args = [IdentifierArgument<"Deallocator", /*opt*/ 1>,
+              ParamIdxArgument<"DeallocatorPtrArgIndex", /*opt*/ 1>];
```
```suggestion
  let Args = [IdentifierArgument<"Deallocator", /*opt=*/1>,
              ParamIdxArgument<"DeallocatorPtrArgIndex", /*opt=*/ 1>];
```

shafik replied:
Nit: Unfortunately we have not been consistent through our codebase in argument comment format but e should strive to be consistent with [bugprone-argument-comment](https://clang.llvm.org/extra/clang-tidy/checks/bugprone/argument-comment.html)

aloisklink replied:
:+1: Fixed in b7e125882fe895f0c4fb346a5ae1acddb70f6739

shafik reviewed (commented):
Just a nit

AaronBallman commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -289,6 +289,11 @@ Bug Fixes to Compiler Builtins
+- Clang now emits a warning instead of an error when using the one or two
+  argument form of GCC 11's ``__attribute__((malloc(deallocator)))``
+  or ``__attribute__((malloc(deallocator, ptr-index)))``
+  (`#51607 <https://github.com/llvm/llvm-project/issues/51607>`).
```
```suggestion
  (`#51607 <https://github.com/llvm/llvm-project/issues/51607>`_).
```

aloisklink replied:
Sorry! That's my fault for being lazy and skipping build the docs! Fixed in 5f6f893c867e73a2a8e04e94a2fa6e2104b9b85e

AaronBallman commented on clang/include/clang/Basic/DiagnosticCommonKinds.td:
```diff
@@ -177,6 +177,10 @@ def warn_unknown_attribute_ignored : Warning<
   "unknown attribute %0 ignored">, InGroup<UnknownAttributes>;
 def warn_attribute_ignored : Warning<"%0 attribute ignored">,
   InGroup<IgnoredAttributes>;
+def warn_multiarg_malloc_attribute_ignored: Warning<
+  "'malloc' attribute ignored because"
+  " Clang does not support the one/two argument form">,
+  InGroup<IgnoredAttributes>;
```
```suggestion
def warn_attribute_form_ignored : Warning<
  "%0 attribute ignored; Clang does not yet support this attribute signature">,
  InGroup<IgnoredAttributes>;
```
If we generalize the wording a bit, we can reuse this diagnostic for other circumstances where we don't yet support a particular signature.

Also, this diagnostic should be moved to DiagnosticSemaKinds.td (it's only diagnosed from `clang/lib/Sema`, so no need for it to be in the "common" diagnostics file for cross-lib diagnostics).

aloisklink replied:
Fixed in 02a153d9be1007b9603bcd14bf3f80d4bf900806, although I've very slightly changed the wording to match other warnings by using `because` instead of the `;`.

I've also added a couple of other error diagnostics in other commit, so please let me know if you think it's worth generalizing them too.

AaronBallman commented on clang/test/Sema/attr-args.c:
```diff
@@ -1,12 +1,14 @@
-inline __attribute__((malloc(a))) void *f5(void);  // expected-error {{'malloc' attribute takes no arguments}}
+inline __declspec(restrict(a)) void *f6_a(void);  // expected-error {{'restrict' attribute takes no arguments}}
+inline __attribute__((malloc(a, 1, a))) void *f6_b(void);  // expected-error {{'malloc' attribute takes no more than 2 arguments}}
+inline __attribute__((malloc(a, 1))) void *f6_c(void);  // expected-warning {{'malloc' attribute ignored because Clang does not support the one/two argument form}}
```
One more test I'd like to see is:
```
inline __attribute__((malloc(1))) void *func(void) // error: '1' does not name a function
```

aloisklink replied:
Fixed in 0659620002757ab8bbd892a717fb47478d22aecd, although with a different error diagnostic. I've also added a bunch more error checking (adapting the code from [`handleCleanupAttr`](https://github.com/llvm/llvm-project/blob/0659620002757ab8bbd892a717fb47478d22aecd/clang/lib/Sema/SemaDeclAttr.cpp#L3793-L3818)), so now we should even be stricter than GCC on what args we accept:

https://github.com/llvm/llvm-project/blob/0659620002757ab8bbd892a717fb47478d22aecd/clang/test/Sema/attr-args.c#L2-L6

https://github.com/llvm/llvm-project/blob/0659620002757ab8bbd892a717fb47478d22aecd/clang/test/Sema/attr-args.c#L13-L21

aloisklink committed:
fixup! [clang] Ignore GCC 11 [[malloc(x)]] attribute

Move diag::warn_multiarg_malloc_attribute_ignored to
`DiagnosticSemaKinds.td` and make it generic so we can use it for other
attributes.

Fixes https://github.com/llvm/llvm-project/pull/68059#discussion_r1355446479

gustedt commented:
I see that the fix is almost ready, good. But generally it would also have helped if the `__has_c_attribute` feature test for this type of borrowed attributes would provide means to distinguish different versions. Here gcc as well as clang only have the value 1. So if the patch would also change the return value for clang to the year whenever the first version of gcc-11 with that feature was released, that would really be helpful.

AaronBallman commented:
> I see that the fix is almost ready, good. But generally it would also have helped if the `__has_c_attribute` feature test for this type of borrowed attributes would provide means to distinguish different versions. Here gcc as well as clang only have the value 1. So if the patch would also change the return value for clang to the year whenever the first version of gcc-11 with that feature was released, that would really be helpful.

Yeah, our current default behavior is that we use the value `1` for almost any vendor-specific attribute. It would be a massive undertaking to try to figure out a consistent date-based value for each of our attributes, and it would be especially complex given that we often have subtle (and sometimes intentional) differences between our implementation and the original implementation. I don't think your suggestion is a bad one, but it's not clear it's workable in practice either. That said, it might be nice for us to have some sort of policy about feature testing changes in semantics of attributes more generally (CC @ErichKeane for awareness).

erichkeane commented:
> > I see that the fix is almost ready, good. But generally it would also have helped if the `__has_c_attribute` feature test for this type of borrowed attributes would provide means to distinguish different versions. Here gcc as well as clang only have the value 1. So if the patch would also change the return value for clang to the year whenever the first version of gcc-11 with that feature was released, that would really be helpful.
> 
> Yeah, our current default behavior is that we use the value `1` for almost any vendor-specific attribute. It would be a massive undertaking to try to figure out a consistent date-based value for each of our attributes, and it would be especially complex given that we often have subtle (and sometimes intentional) differences between our implementation and the original implementation. I don't think your suggestion is a bad one, but it's not clear it's workable in practice either. That said, it might be nice for us to have some sort of policy about feature testing changes in semantics of attributes more generally (CC @erichkeane for awareness).

I'd actually been thinking about this all morning, and I don't see how we could do it in a way that wouldn't make the world a worse place.  The reason feature-test-macros work is because there is an agreed upon 'source' of those values, without Clang and GCC agreeing on a value here (which we do in C++ via SD10), anything we do is going to be no better than just compiler version checks unfortunately.

gustedt commented:
Hi, 
yes the really bad choice here is by gcc to have the same name for basically two different attributes.
For the value, they also missed the opportunity to do something sensible when moving to C++ attributes, what a pitty. 

For a concrete guideline in that jungle, when (= which version) do you expect this patch to hit distribution? Will it be in clang-18 once that is released?

Can we expect a similar feature, let's Call it clang::deallocator_function in a future version?
Thanks
 Jens
-- 
Jens Gustedt - INRIA & ICube, Strasbourg, France

erichkeane commented:
> Hi, yes the really bad choice here is by gcc to have the same name for basically two different attributes. For the value, they also missed the opportunity to do something sensible when moving to C++ attributes, what a pitty. For a concrete guideline in that jungle, when (= which version) do you expect this patch to hit distribution? Will it be in clang-18 once that is released? Can we expect a similar feature, let's Call it clang::deallocator_function in a future version? Thanks Jens -- Jens Gustedt - INRIA & ICube, Strasbourg, France

It does appear that this is on track for clang 18, but it hasn't finished review yet, so I cannot promise that of course.  At the moment, nothing like a `clang::deallocator_function` is proposed, so unless someone does the work to implement and propose it, we don't have a plan to do so.

aloisklink committed:
fixup! [clang] Ignore GCC 11 [[malloc(x)]] attribute

Support bugprone-argument-comment format.

Co-authored-by: Shafik Yaghmour <shafik@users.noreply.github.com>

aloisklink committed:
fixup! [clang] Ignore GCC 11 [[malloc(x)]] attribute

Fix incorrect RST link syntax.

Co-authored-by: Aaron Ballman <aaron@aaronballman.com>

aloisklink committed:
fixup! [clang] Ignore GCC 11 [[malloc(x)]] attribute

Error if `deallocator` and `ptrIdx` to
`[[malloc(deallocator, ptrIdx)]]` have invalid types.

aloisklink commented:
I've fixed reviewer comments! Sorry for the delay! I didn't have much time, and my PC isn't the fastest, so building Clang + regression tests takes a while!

As recommended by https://github.com/llvm/llvm-project/pull/68059#discussion_r1355449108, I added type checking for the attribute arguments so that we should throw an error on any invalid args that GCC throws an error (or even a warning on). Since this is a pretty substantial change, I made it all as `fixup!` commits to make the changes easier to review.

---

> gcc to have the same name for basically two different attributes

I agree, it's not ideal! Having a separate name for this attribute (e.g. `[[dealloc(func)]]` or `[[deallocator(func)]]` would be my preference. Although I think it might be worth asking the GCC team to implement it first. After all, this PR doesn't actually add support for the `[[malloc(deallocator)]]` attribute form in Clang, it just ignores it, instead of throwing an error.

erichkeane commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -2064,13 +2064,87 @@ static void handleTLSModelAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+    }
+  }
+
+  // FIXME: we should add this attribute to Clang's AST, so that clang-analyzer
```
I think we SHOULD still add this to the AST anyway, as it represents code still.  However we should teach `CGCall.cpp` to no longer emit the 'no-alias'.

erichkeane replied:
Note this would require a codegen test!

erichkeane reviewed (commented):
Note that this just came up in another bug report (see discussion here: https://github.com/llvm/llvm-project/issues/128026).

it seems we let this go, so I was hoping we could get this merged!.  

@aloisklink : I see you're still around github, but seem to have moved onto other projects.  If you're still interested in working on this, please make the changes suggested and we can merge this.  If you're not, please let me know and I'll commander this.

Sorry for the delay!

erichkeane commented:
also-also: A bug report to improve the diagnostic for `-Wmismatched-dealloc` based on this being in the AST would be appreciated.  If you could file one, it would be greatly appreciated.

Again, sorry for the delay on this one, we managed to drop the ball on this review.

erichkeane commented:
Closing, as I separately submitted this with my changes here: [79a28aa](https://github.com/llvm/llvm-project/commit/79a28aa0a48feba34ddc3c1791ea0be88f354542)

Note original author should be in tact.  Thank you for your contribution!

erichkeane commented:
> This is linked to #159080

See #159371
